Обновление приложения для оценки фильмов

Elite-tea opened the pull request:
feat: Добавлена возможность добавить пользователя в друзья и удалить
feat: Добавлена возможность выводить общих друзей
feat: Добавлена возможность ставить и удалять лайки к фильмам
feat: Добавлена возможность выводить топ фильмов по лайкам, при этом самому указывать сколько фильмов будет в топе
feat: Добавлены тесты для пользователя
fix: Доработаны проверки при создании пользователей и обновлении
refactor: Переработан метод валидации для пользователя
style: Код приведен к общему стилю практикума

Elite-tea committed:
feat: Добавлена возможность добавить пользователя в друзья и удалить
feat: Добавлена возможность выводить общих друзей
feat: Добавлена возможность ставить и удалять лайки к фильмам
feat: Добавлена возможность выводить топ фильмов по лайкам, при этом самому указывать сколько фильмов будет в топе
feat: Добавлены тесты для пользователя
fix: Доработаны проверки при создании пользователей и обновлении
refactor: Переработан метод валидации для пользователя
style: Код приведен к общему стилю практикума

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,90 @@
+            return films.get(id);
+        } else {
+            log.debug("Фильм не существует");
+            throw new NotFoundException(String.format("Фильм с id %s не существует", films.get(id)));
```
Если фильм не существует, то films.get(id) вернет null - вам здесь только id нужен, иначе у вас сообщение будет неинформативным (а в какой-то степени даже вредным).

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,90 @@
+    public User getByIdUser(Long id) {
+        if (users.containsKey(id)) {
+            log.debug("Запрошен пользователь: {}", users.get(id));
+            return users.get(id);
```
Вы дважды ходите в хранилище, чтобы получить пользователя:
1. для логирования
2. для возврата
Если у вас пользователи будут на удаленном сервере в другой части света, работа метода будет занимать много времени.

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,67 +1,58 @@
 @RequestMapping("/films")
-@Slf4j
+@AllArgsConstructor
 public class FilmController {
```
У вас в одним классах очень хорошее документирование кода, в других даже строчки нет. Хотелось бы везде красиво сделать.

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,67 +1,58 @@
-     * @param film информация о фильме.
-     */
+    @Autowired
+    private final FilmStorage filmStorage;
```
Есть мнение, что хранилище не должно быть доступно из контроллера, а должно вызываться из сервиса. Это, в первую очередь, помогает добавиться слабой связанности кода, что облегчает доработки в будущем.

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,76 @@
+        if (userStorage.getByIdUser(userId) != null) {
+            if (filmStorage.getByIdFilm(filmId) != null) {
+                filmStorage.getByIdFilm(filmId).addLike(userId);
+
```
Такую лесенку операторов часто называют callback hell (пришло из javascript). Такое делать не возбраняется, но усложняет читать. Некоторые разработчики делают обратную проверку, отчего код "выравнивается". Пример:

```
if (userStorage.getByIdUser(userId) == null) {
    throw new Exception()
}
if (filmStorage.getByIdFilm(filmId) == null) {
    throw new Exception2()
}
filmStorage.getByIdFilm(filmId).addLike(userId);
```

То же самое, но читаемость улучшена.

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,76 @@
+     */
+
+    public void deleteLike(Long userId, Long filmId) {
+        if (userStorage.getByIdUser(userId) != null) {
```
Тут также можно поправить, как и выше.

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,76 @@
+     */
+
+    public List<Film> getPopularFilm(int topNumber) {
+        return filmStorage.getFilm().stream().sorted(Comparator.comparingInt(Film::getLike).reversed()).limit(topNumber).collect(Collectors.toList());
```
Стримы для лучшей читаемости я бы советовал разбивать по операциям:

```
list.stream()
     .filter()
     .map()
     .collect();
```

Elite-tea committed:
fix: Оптимизирована работы вывода логов, теперь вместо целых объектов, отображаются их id используемые для запросов
fix: Изменена логика доступа к хранилищу, теперь доступ реализован через сервис объекты
style: Улучшено чтение кода, посредством добавление отступов и пробелов
refactor: Дополнена документация, теперь описывается каждый метод, класс и поле, за исключение тестов

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,67 +1,101 @@
-            throw new ValidationException("Данного фильма нет в базе данных");
-        }
-        return film;
+        return filmService.getFilmStorage().put(film);
```
Неплохо, но я имел немного другую форму. В сервисе создается метод (н-р, addFilm()), в котором вызывается filmStorange.put(). Почему за такую форму выступают? В этом случае контроллер ничего не знает хранилище. Представьте, что у вас один сервис будет работать с мапой в качестве хранилища, а другой с базой данных. В вашем случае придется вносить изменения в метод контроллера, потому что сервис для работы с базой данных не будет иметь метод put(). При предлагаемом мною подходе контроллеру вообще не интересно, как устроен метод addFilm() в сервисе - он его просто вызывает. А у вас по какому-нибудь параметру будет меняться сервис с мапой на сервис с базой данных. Надеюсь, понятно объяснил.

Elite-tea committed:
fix: Изменена логика доступа к хранилищу, теперь доступ реализован через сервис(прокладку)
refactor: Дополнена документация для новых методов

Elite-tea committed:
style: Добавлены недостающие отступы для улучшения чтения кода

kriwsha commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,129 @@
+     * @param film объект фильма
+     * @return возвращает созданный фильм
+     */
+    public Film addFilms(@Valid @RequestBody Film film) {
```
В сервисе уже не нужно добавлять эти аннотации. Это все должно остаться в контроллере.  В сервис попадает уже только сам объект. Давайте удалим эти аннотации здесь и в других методах.

Elite-tea committed:
fix: Удалены аннотации для работы контроллера в сервис клаасе film
